### ISSUE TITLE
Root Tenant to read & write Entites of other Tenants

### ISSUE THREAD

**Issue by eluvitie** [Finbuckle/Finbuckle.MultiTenant#593]
Hi There

I have the following use case and I'm stuck on this problem...

**Use Case**
In my application there is one root Tenant which has special rights to view and modify some of the entites of the other tenants. In the application there are Products which are not restricted by the `.IsMultiTenant()` constraint. And then there is a `BasketItem` Entity which each Tenant only sees their own `BasketItems`.

**Goal**
The Users of the Root Tenant should be able to add products to the Basket of another tenant.

**What I’ve tried so far**

- Listing Basket Entites from other Tenants works by using the Query Option ` .IgnoreQueryFilters()`

- I’ve added the `TenantId` property to the `BasketItem` Entity so I can set it manually if needed.

- Then I got the “1 modified entities with Tenant Id mismatch.” Error, when I tried to add a new item with the root Tenant to another tenant.

- According to the [documentation](https://www.finbuckle.com/MultiTenant/Docs/v6.1.0/EFCore#tenant-mismatch-mode) I should be able to set the Mismatch Mode to TenantMismatchMode.Ignore and then it should allow me to set the tenantId manually

- In the DbContext Class I tried to set this option:
```
public class DbContext : MultiTenantIdentityDbContext
{
    ....
    public TenantMismatchMode TenantMismatchMode { get; set; } = TenantMismatchMode.Ignore;
}
```

→ Still the same “TenantId Mismatch Error”

- I also tried to set MyTenantInfo Class manually to the target tenant, but this doesn’t work. I assume because the tenant will be set in the `ITenantInfo` via the Strategy (for example the `WithClaimStrategy()`)

```
public class MyTenantInfo : ITenantInfo
{
    public MyTenantInfo(
        string id,
        string name,
        string? connectionString,
    )
    {
        Id = id;
        Identifier = id;
        Name = name;
        ConnectionString = connectionString ?? string.Empty;
    }

    
    public string Id { get; set; } = default!;
    public string Identifier { get; set; } = default!;
    public string Name { get; set; } = default!;
    public string ConnectionString { get; set; } = default!;
  }
```

**Questions**
- What is the best way to allow users of a root tenant to view, add, update, delete entites of the other tenants?
- Is there a way to set the ITenantInfo manually, so I could change it before the database operation will be called?

Thank you for you help!

**Comment by AndrewTriesToCode** [comment]
hi @eluvitie sorry for the slow reply. I need some time to check the `TenantMisMatchMode` because that would have been my first suggestion.

My next suggestion is to "new" up a dbcontext for the specific tenant you want to enter items for, then dispose of it. I'm working on convenience factory methods that will help with this but for not doing it manually isn't too bad, but it is not very efficient in some use cases.

**Comment by AndrewTriesToCode** [comment]
One more thing, rather than this:
```cs
public TenantMismatchMode TenantMismatchMode { get; set; } = TenantMismatchMode.Ignore;
```

in your constructor do this:
```cs
this.TenantMismatchMode = TenantMisMatchMode.Ignore;
```

or when you use the dbcontext you can set it like this right before you call `SaveChanges`:
```cs
myDbContext.TenantMisMatchMode.Ignore;
```

I suspect your version is creating a new property that the base implementation isn't seeing. Since you are inheriting from `MultiTenantDbContext` it already has the property for you to use. If you were instead implementing `IMultiTenantDbContext` your approach would be needed since there would be no base class.

**Comment by eluvitie** [comment]
Hi @AndrewTriesToCode 
Thank you for your response. 
Both of your approached worked for me. 

I went for the latter, since I think its best to only change the `TenantMisMatchMode` if needed and not change it globally.

In the project I use the `Ardalis.Specification` Package, so I extended my `IRepository` Interface with a function `setTenantMismatchMode(TenantMismatchMode mode)` 


This function changes the default `TenantMismatchMode` to what ever is desired in the DbContext. In my Case I'm setting it to `TenantMismatchMode.Ignore`.


```
DbRepository : IRepository
{
....
    public void setTenantMismatchMode(TenantMismatchMode mode)
    {
       this._dbContext.TenantMismatchMode = mode;
    }
....
}
```

In the Handler I then call this before I'm making changes to the DB: `_repository.setTenantMismatchMode(TenantMismatchMode.Ignore); `

Maybe this helps someone else. And if there are better ways to do it, I'm open for new approaches.
